### DISCUSSION TITLE
Loop the video with SetFFmpegVideoFilters is failing

### DISCUSSION THREAD

**Discussion by Dush27** [ffmpeginteropx/FFmpegInteropX#304] [Q&A]
Hi Team,

I am trying to loop the video by using below code, but for some reason the looping option seems to fail.

**ffmpegMediaSource2_.SetFFmpegVideoFilters(L"loop=loop=-1,crop=322:181:0:149,scale=640:360");**

The same  looping issue is not observed if I don't use above line of code to crop and scale the video.

Can you please provide your suggestions ?

NOTE: My intention is to crop, scale and perform loop on the video.

Thanks,
Dushyanth

**Comment by lukasf** [comment]
Hi @Dush27,

for the loop filter, you need to specify all three options:

This would loop the first frame infinitely:
loop=loop=-1:size=1:start=0

If you want to loop the whole video, you need to know the number of frames contained. I can use this with or without scaling. What troubles me is that I see skyrocketing memory usage. Not sure yet what is going on.

Other approach: How about using MediaPlayer.IsLoopingEnabled instead of a filter? The loop filter is generally memory intensive. Looping on MediaPlayer would be the preferred approach. Also, MediaPlayer does not like it when the file plays longer than the duration we set.

**Reply by Dush27** [reply]
Hi @lukasf ,

Thanks for your reply.

I tried using the filter as below as the number of frames in the video is 60. Now I see that video itself is not getting played.
Also, as you said, I too observed the memory getting increased as lot while using loop.

**loop=loop=-1:size=60:start=0,crop=322:181:0:149,scale=640:360**

Also, w.r.t MediaPlayer.IsLoopingEnabled, I cannot use it here as the requirement is to support video as virtual background with various video formats.

Thanks,
Dushyanth

**Reply by lukasf** [reply]
That's strange, it works for me. Can you you upload a sample file?

> Also, w.r.t MediaPlayer.IsLoopingEnabled, I cannot use it here as the requirement is to support video as virtual background with various video formats.

And why not use MediaPlayer for that? Put MediaPlayerElement into background, set AreTransportControlsEnabled=false and assign MediaPlayer with IsLoopingEnabled=true.

**Reply by Dush27** [reply]
https://user-images.githubusercontent.com/114069851/192265410-30451bb8-093c-413e-b5ce-7cdd923b0d1f.mp4

Attached the sample as requested by you.

Also, regarding mediaPlayer, I was trying to code as below.

                ffmpegMediaSource2_.SetFFmpegVideoFilters(wss_.str().c_str());
                winrt::Windows::Media::Playback::MediaPlaybackSession playbackSession = ffmpegMediaSource2_.PlaybackSession();
                playbackSession.MediaPlayer().IsLoopingEnabled(true);

I am facing an error with above code, though I am including the corresponding header file. Looks like MediaPlayer is not defined as part of this library ? Correct me if i am missing something here ?

1>C:\Users\Dushyanth\rcvrnd\modules\nvbg\win\video_file_playback.cpp(766,32): error C3779: 'winrt::impl::consume_Windows_Media_Playback_IMediaPlaybackSession<winrt::Windows::Media::Playback::IMediaPlaybackSession>::MediaPlayer': a function that returns 'auto' cannot be used before it is defined
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\cppwinrt\winrt\impl\Windows.Media.Playback.0.h(2000): message : see declaration of 'winrt::impl::consume_Windows_Media_Playback_IMediaPlaybackSession<winrt::Windows::Media::Playback::IMediaPlaybackSession>::MediaPlayer'

Thanks,
Dushyanth

**Reply by lukasf** [reply]
The sample file plays fine for me and loop filter works in general. I guess that there is a different problem in your app. But as I said, MediaPlayer does not like it when the file plays longer than the expected duration. So after the file played once at normal speed, all following frames are displayed way too fast, probably because the MediaPlayer tries to reach the end as soon as possible

But if I set IsLoopingEnabled on MediaPlayer, instead of using the loop filter, it loops cleanly as expected.

Please check our CPP or CS sample to see how to use MediaPlayer. You create one instance, e.g. in the constructor of your page or in your ViewModel or whatever, you have a MediaPlayerElement on your page, and in the page constructor, you call mediaPlayerElement.SetMediaPlayer(mediaPlayer) to assign the player to your MediaPlayerElement.

Then each time you want to switch to a different file, do it like this:
```
// Create FFmpegMediaSource
FFmpegMSS = co_await FFmpegMediaSource::CreateFromStreamAsync(stream, Config);
playbackItem = FFmpegMSS->CreateMediaPlaybackItem();

// Pass MediaPlaybackItem to MediaPlayer
mediaPlayer->Source = playbackItem;
```

If I edit the sample app, put IsLoopingEnabled = true in the MainPage contstructor and open your file, it loops as expected. Just try it and see how it works in the sample. Be sure to store the FFmpegMediaSource in a field, so it is kept in memory as long as the file played.

**Reply by Dush27** [reply]
Hi Lukas,

In my requirement, I cannot use MediaPlayer as I am not decoding and sending it to the UI.
Instead, I should actually decode and send the raw buffer to the encoder. 

Is there any other way inside FFmpegInteropX, to seek to the start of the file when I reach to the end of the file?

Thanks,
Dushyanth

**Reply by lukasf** [reply]
May I ask how you do the encoding? And where do you send it, do you create a live stream to the target device or app? If you can describe your setup, I can better help you on how you can make use of FFmpegInteropX.

Depending on your scenario, you might be able to use our FrameGrabber, maybe a bit like this:


```
            auto frameGrabber = co_await FrameGrabber::CreateFromStreamAsync(stream);
            while (true)
            {
                auto frame = co_await frameGrabber->ExtractNextVideoFrameAsync();
                if (!frame)
                {
                    // frame is null on end of file. skip back to beginning.
                    frame = co_await frameGrabber->ExtractVideoFrameAsync(TimeSpan{ 0 });
                }

                if (frame)
                {
                    // send to encoder
                }
                else
                {
                    // abort if no frame received at all
                    break;
                }
            }
```

The VideoFrame contains raw bitmap data, size and timestamp. You would have to calculate increasing timestamps for the encoder.

**Reply by brabebhin** [reply]
Mediaplayer has some support for streaming and casting.

**Reply by Dush27** [reply]
Hi Lukasf,

I am using FFmpegInteropX as below where I want to decode the file and read the samples. Each sample is read in an regular interval which is passed to application to use it as background at same regular interval.

a) Used FFmpegMediaSource::CreateFromUriAsync and wait_for to get FFmpegInteropXWinUI::FFmpegMediaSource object
b) Used FFmpegMediaSource::GetMediaStreamSource() and convert to IMFGetService. From IMFGetService fetch MF_MEDIASOURCE_SERVICE object
c) Used MFCreateSourceReaderFromMediaSource with IMFMediaSource, also in attributes have set MF_SOURCE_READER_D3D_MANAGER, MF_READWRITE_ENABLE_HARDWARE_TRANSFORMS = true and MF_SOURCE_READER_ENABLE_ADVANCED_VIDEO_PROCESSING = true
d) Also setting some filters to crop and scale using ffmpegMediaSource2_.SetFFmpegVideoFilters(wss_.str().c_str()) .

The problem here is that when Read Sample reaches ENDOFSTREAM, I am not able to loopback.

        winrt::check_hresult(reader_->ReadSample(
            MF_SOURCE_READER_FIRST_VIDEO_STREAM,
            0,
            nullptr,
            &flags,
            &timeStamp,
            sample.put()));

        if (flags & MF_SOURCE_READERF_ENDOFSTREAM || sample == nullptr) {
            //I am looking for a way to loop the video when it reaches the ENDOFSTREAM.
        }

**NOTE: I had a look at the suggestion of  using FrameGrabber. But it seems like MediaStreamSource and FrameGrabber are two different items.** 

Thanks,
Dushyanth

**Reply by lukasf** [reply]
Just do SetCurrentPosition(0) on IMFSourceReader on file end. Like this:

```
HRESULT SetPosition(IMFSourceReader *pReader, const LONGLONG& hnsPosition)
{
    PROPVARIANT var;
    HRESULT hr = InitPropVariantFromInt64(hnsPosition, &var);
    if (SUCCEEDED(hr))
    {
        hr = pReader->SetCurrentPosition(GUID_NULL, var);
        PropVariantClear(&var);
    }
    return hr;
}
```

FrameGrabber generally could be used as well, but it does software decoding only. For your scenario, IMFSourceReader might be better.

**Reply by Dush27** [reply]
Hi Lukas,

-> I have tried this SetPosition already which works to loop back to start of the file when I don't use SetFFmpegVideoFilters(). 
     Point to be noted is that there is lag(approx 2-3seconds delay for a 20seconds playback) everytime when ReadSample is 
     performed for first sample when end of stream is reached.Can we avoid this lag by some means? I guess it is the 
     processing of the file again by ffmpeg?

**FFmpegInteropX log without SetFFmpegVideoFilters():**

```

GetNextSample
 - QueuePacket
 - PopPacket
GetNextSample
GetNextPacket reaching EOF
End of stream reached. Enter draining mode.
End of stream reached. No more samples in decoder.
End of stream reached.
DisableStream
Flush
EnableStream
Flush
GetNextSample
 - QueuePacket
 - PopPacket
GetNextSample_

```

-> If I use SetPosition with SetFFmpegVideoFilters(), loop back to start of the file fails and video is stuck at the last frame.

**FFmpegInteropX log with SetFFmpegVideoFilters():**

```
_GetNextSample
 - QueuePacket
 - PopPacket
GetNextSample
GetNextPacket reaching EOF
End of stream reached. Enter draining mode.
End of stream reached. No more samples in decoder.
End of stream reached.
DisableStream
Flush
EnableStream
Flush
GetNextSample
End of stream reached. No more samples in decoder.
End of stream reached.
DisableStream
Flush
EnableStream
Flush
GetNextSample
End of stream reached. No more samples in decoder.
End of stream reached.
DisableStream
Flush
EnableStream_
```

**NOTE:** _**I have used following code to avoid the issue with Filters but i am not able to avoid the delay for the first ReadSample() invocation when the end of stream is reached.**_

```
if (flags & MF_SOURCE_READERF_ENDOFSTREAM || sample == nullptr) {
            //Below change is a workaround to loop the video as FFmpegInteropX loop fails 
            //when we apply crop and scale using SetFFmpegVideoFilters
            **ffmpegMediaSource2_.DisableVideoEffects();**
            if (SourceCanSeek()) {
                 winrt::check_hresult(SetPosition(LONGLONG(0)));
                 **ffmpegMediaSource2_.SetFFmpegVideoFilters(wss_.str().c_str());**
            } 
}
```
Thanks,
Dushyanth

**Reply by lukasf** [reply]
Thanks for reporting the issue with filters on EOF. I have a fix, please try branch filters-eof.

I don't know where the lag comes from. If I seek to beginning with normal MediaPlayer and start playback, it happens nearly instantly. So maybe it has to do with IFMSourceReader.

**Reply by lukasf** [reply]
One alternative would be using MediaPlayer in frame server mode:

https://learn.microsoft.com/en-us/windows/uwp/audio-video-camera/play-audio-and-video-with-mediaplayer#use-mediaplayer-in-frame-server-mode

You have full HW acceleration and get an event when a new frame is to be rendered. You copy the frame to a surface and forward that to your encoder or whatever you have behind the decoder part. You don't need to care about timestamps because MediaPlayer raises the events in the right moment. Then you can set IsLoopingEnabled=true on the MediaPlayer and it will do smooth and fast looping of the video file. The file you sent plays without any noticeable delay with IsLoopingEnabled.

See DirectXInteropHelper and D3D11VideoSampleProvider on how to interop from IDirect3DSurface <-> IDXGISurface <-> ID3D11Texture2D. You do not need to use Win2D as in the docs. You can use any ID3D11Texture2D as target, I think only ArraySize needs to be 1, otherwise you cannot create a IDirect3DSurface from it.

**Reply by Dush27** [reply]
Hi Lukas,

Sorry for the late response.

I gave a try with "filters-eof" and below is my observation.
   -> I don't see the video playback fail while reaching the end of stream, but the filters which are applied while creating the reader is not getting applied automatically after reaching the end of stream.

  I had to apply the filters again while performing SetPosition as below, to apply the filters while reaching the end of the stream as below.

```
if (flags & MF_SOURCE_READERF_ENDOFSTREAM || sample == nullptr) {
            if (SourceCanSeek()) {
                winrt::check_hresult(SetPosition(LONGLONG(0)));
                ffmpegMediaSource2_.SetFFmpegVideoFilters(wss_.str().c_str());
             }
 }
```
**Regarding your suggestion of MediaPlayer is frame server mode, can we follow this method to integrate with win32 application ?**

Thanks,
Dushyanth

**Reply by lukasf** [reply]
Hi @Dush27, yes frame server mode definitely works in Win32 Desktop apps.

You are right, the fix in filters-eof did not really work. It kept playing, but it silently dropped the filter effect. I read a bit about filters, and it seems that we should not even try to drain them, because they won't be usable after that. I have updated the branch with a new fix, and it looks like this one really works and keeps the filter applied after eof. This one also works with MediaPlayer.IsLoopingEnabled=true.

**Comment by Dush27** [comment]
Hi @lukasf,

Now the filter branch seems to work fine.

Thanks,
Dushyanth

**Reply by lukasf** [reply]
Good to hear that.
